src/fold.rs

```diff
@@ -500,39 +502,37 @@ fn shift_right(
 fn shift_left(
     left: Expr,
     right: Expr,
-    ctype: &Type,
+    _ctype: &Type,
     location: &Location,
 ) -> CompileResult<ExprType> {
     let (left, right) = (left.const_fold()?, right.const_fold()?);
     if let ExprType::Literal(token) = right.expr {
         let shift = match token.non_negative_int() {
             Ok(u) => u,
-            Err(_) => semantic_err!("cannot shift left by a negative amount".into(), *location),
+            Err(_) => {
+                return Err(location.error(SemanticError::NegativeShift { is_left: true }));
+            }
         };
+
         if left.ctype.is_signed() {
             let size = match left.ctype.sizeof() {
                 Ok(s) => s,
                 Err(err) => semantic_err!(err.into(), *location),
             };
             let max_shift = u64::from(CHAR_BIT) * size;
             if shift >= max_shift {
-                semantic_err!(
-                    format!(
-                        "cannot shift left by {} or more bits for type '{}' (got {})",
-                        max_shift, ctype, shift
-                    ),
-                    *location,
-                );
+                return Err(location.error(SemanticError::TooManyShiftBits {
+                    is_left: true,
+                    current: shift,
```

Comment:
This doesn't print the type any more. I'm not tied to that, but is there a reason you removed it?